src/libraries/Microsoft.Extensions.Options.DataAnnotations/src/DataAnnotationValidateOptions.cs

```diff
@@ -52,19 +55,86 @@ public ValidateOptionsResult Validate(string? name, TOptions options)
+            "statically analyzed so its members may be trimmed.")]
+        private static bool TryValidateOptions(object options, string qualifiedName, List<ValidationResult> results, ref List<string>? errors, ref HashSet<object>? visited)
+        {
+            if (visited is not null && visited.Contains(options))
             {
-                errors.Add($"DataAnnotation validation failed for '{typeName}' members: '{string.Join(",", result.MemberNames)}' with the error: '{result.ErrorMessage}'.");
+                return true;
             }
 
-            return ValidateOptionsResult.Fail(errors);
+            results?.Clear();
+
+            bool res = Validator.TryValidateObject(options, new ValidationContext(options), results, validateAllProperties: true);
+            if (!res)
+            {
+                errors ??= new List<string>();
+
+                foreach (ValidationResult result in results!)
+                {
+                    errors.Add($"DataAnnotation validation failed for '{qualifiedName}' members: '{string.Join(",", result.MemberNames)}' with the error: '{result.ErrorMessage}'.");
+                }
+            }
+
+            foreach (PropertyInfo propertyInfo in options.GetType().GetProperties(BindingFlags.Instance | BindingFlags.Public))
+            {
+                if (propertyInfo.GetMethod is null)
+                {
+                    continue;
+                }
+
+                object? value = propertyInfo!.GetValue(options);
+
+                if (value is null)
+                {
+                    continue;
+                }
+
+                if (propertyInfo.GetCustomAttribute<ValidateObjectMembersAttribute>() is not null)
+                {
+                    visited ??= new HashSet<object>(new ReferenceComparer());
+                    visited.Add(options);
+
+                    results ??= new List<ValidationResult>();
+                    res = TryValidateOptions(value, $"{qualifiedName}.{propertyInfo.Name}", results, ref errors, ref visited) && res;
+                }
+                else if (propertyInfo.GetCustomAttribute<ValidateEnumeratedItemsAttribute>() is not null &&
+                        value is IEnumerable enumerable)
+                {
+                    visited ??= new HashSet<object>(new ReferenceComparer());
+                    visited.Add(options);
+                    results ??= new List<ValidationResult>();
+
+                    int index = 0;
+                    foreach (object item in enumerable)
+                    {
+                        res = TryValidateOptions(item, $"{qualifiedName}.{propertyInfo.Name}[{index++}]", results, ref errors, ref visited) && res;
+                    }
+                }
+            }
+
+            return res;
+        }
+
+        private readonly struct ReferenceComparer : IEqualityComparer<object>
```

Comment:
Why make this a struct? The two places it's being used, it'll be immediately boxed.

Also, on .NET 5+ this isn't needed; the HashSets can be constructed passing in `ReferenceEqualityComparer.Instance`.  You can just make this type here a polyfill with the same shape, and then just compile this in downlevel.